File: src/main/kotlin/com/huawei/cangjie1/psi/CjModifierListOwnerStub.java
package com.huawei.cangjie1.psi;

import com.huawei.cangjie1.lexer.CjModifierKeywordToken;
import com.huawei.cangjie1.psi.psiUtil.AddRemoveModifierKt;
import com.intellij.lang.ASTNode;
import com.intellij.psi.stubs.IStubElementType;
import com.intellij.psi.stubs.StubElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import com.huawei.cangjie1.psi.stubs.elements.CjStubElementTypes;
import java.util.List;


public class CjModifierListOwnerStub<T extends StubElement<?>> extends CjElementImplStub<T> implements CjModifierListOwner {
    public CjModifierListOwnerStub(ASTNode node) {
        super(node);
    }

    public CjModifierListOwnerStub(T stub, IStubElementType nodeType) {
        super(stub, nodeType);
    }

    @Override
    @Nullable
    public CjModifierList getModifierList() {
        return getStubOrPsiChild(CjStubElementTypes.MODIFIER_LIST);
    }

    @Override
    public boolean hasModifier(@NotNull CjModifierKeywordToken modifier) {
        CjModifierList modifierList = getModifierList();
        return modifierList != null && modifierList.hasModifier(modifier);
    }

    @Override
    public void addModifier(@NotNull CjModifierKeywordToken modifier) {
        AddRemoveModifierKt.addModifier(this, modifier);
    }

    @Override
    public void removeModifier(@NotNull CjModifierKeywordToken modifier) {
        AddRemoveModifierKt.removeModifier(this, modifier);
    }





}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjModifierListOwnerStub.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.descriptors.DescriptorVisibilities
import com.linqingying.cangjie.descriptors.DescriptorVisibility
import com.linqingying.cangjie.lexer.CjModifierKeywordToken
import com.linqingying.cangjie.psi.psiUtil.addModifier
import com.linqingying.cangjie.psi.psiUtil.removeModifier
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes
import com.linqingying.cangjie.resolve.ModifiersChecker.Companion.resolveVisibilityFromModifiers
import com.intellij.lang.ASTNode
import com.intellij.psi.stubs.IStubElementType
import com.intellij.psi.stubs.StubElement
import com.linqingying.cangjie.lexer.CjKeywordToken

open class CjModifierListOwnerStub<T : StubElement<*> > : CjElementImplStub<T>,
    CjModifierListOwner {
    constructor(node: ASTNode) : super(node)

    constructor(stub: T, nodeType: IStubElementType<*, *>) : super(stub, nodeType)

    override val modifierList: CjModifierList?
        get() = getStubOrPsiChild(CjStubElementTypes.MODIFIER_LIST)

    override fun hasModifier(modifier: CjKeywordToken): Boolean {
        val modifierList = modifierList
        return modifierList != null && modifierList.hasModifier(modifier)
    }

    override val annotations: List<CjAnnotation>
        get() {
            val modifierList = modifierList ?: return emptyList()
            return modifierList.annotations
        }

    override val annotationEntries: List<CjAnnotationEntry>
        get() {
            val modifierList = modifierList ?: return emptyList()
            return modifierList.annotationEntries
        }
    override val modifierVisibility: DescriptorVisibility
        get() = resolveVisibilityFromModifiers(
            this,
            DescriptorVisibilities.INTERNAL
        )

    override fun addModifier(modifier: CjKeywordToken) {
        addModifier(this, modifier)
    }

    override fun removeModifier(modifier: CjKeywordToken) {
        removeModifier(this, modifier)
    }
}
